src/frontend/src/heartbeat.rs

```diff
@@ -79,27 +104,60 @@ impl HeartbeatTask {
         });
     }
 
-    fn start_heartbeat_report(&self, req_sender: HeartbeatSender) {
+    fn start_heartbeat_report(
+        &self,
+        req_sender: HeartbeatSender,
+        mut outgoing_rx: Receiver<OutgoingMessage>,
+    ) {
         let report_interval = self.report_interval;
 
         common_runtime::spawn_bg(async move {
+            let sleep = tokio::time::sleep(Duration::from_millis(0));
+            tokio::pin!(sleep);
+
             loop {
-                let req = HeartbeatRequest::default();
+                let req = tokio::select! {
+                    message = outgoing_rx.recv() => {
+                        if let Some(message) = message {
+                            match outgoing_message_to_mailbox_message(message) {
+                                Ok(message) => {
+                                    let req = HeartbeatRequest {
+                                        mailbox_message: Some(message),
+                                        ..Default::default()
+                                    };
+                                    Some(req)
+                                }
+                                Err(e) => {
+                                    error!(e;"Failed to encode mailbox messages!");
+                                    None
+                                }
+                            }
+                        } else {
+                            None
```

Comment:
I think we should break here. The impl of heartbeat in frontend is slightly different than that in datanode. When retrying heartbeat in frontend, it starts from creating a clean new mailbox, and re-spawn this bg task. The `outgoing_rx` is closed, `recv` always returns `None`, but the old task is not stopped until you break here. This could effectively create a endless loop.